src/main/java/org/testng/Assert.java

```diff
@@ -152,7 +152,7 @@ private static boolean areNotEqualImpl(Object actual, Object expected) {
     if ((expected == null) && (actual == null)) {
       return false;
     }
-    if (expected == null ^ actual == null) {
+    if(((expected == null) && !actual.equals(expected)) ^ ((actual == null) && !expected.equals(actual))){
       return true;
     }
```

Comment:
This might cause a `NullPointerException` in line 158 when `equals(...)` erroneously returns `true` for `null` here.
Maybe the following would be better:
```suggestion
    if (expected == null) {
      return !actual.equals(expected);
    }
    if (actual == null) {
      return !expected.equals(actual);
    }
```